### PR TITLE
HHH 15454

### PR DESCRIPTION
**Pull request by karge-itestra**
https://hibernate.atlassian.net/browse/HHH-15454

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java [hunk 1]
```diff
@@ -100,6 +101,9 @@ public Object get(String alias) {
 
 		public <X> X get(String alias, Class<X> type) {
 			final Object untyped = get( alias );
+			if (type.isPrimitive() && untyped != null && !untyped.getClass().isPrimitive()) {
+				type = PrimitiveWrapperHelper.getDescriptorByPrimitiveType(type).getWrapperClass();
+			}
 			if ( untyped != null ) {
 				if ( !type.isInstance( untyped ) ) {
 					throw new IllegalArgumentException(
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java:109]:
```suggestion
			if ( untyped != null ) {
				if ( !type.isInstance( untyped ) && ( !type.isPrimitive()  || !PrimitiveWrapperHelper.getDescriptorByPrimitiveType(type).getWrapperClass().isInstance( untyped ) ) {
					throw new IllegalArgumentException(
```

File: hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java [hunk 2]
```diff
@@ -112,7 +116,7 @@ public <X> X get(String alias, Class<X> type) {
 					);
 				}
 			}
-			return (X) untyped;
+			return type.cast(untyped);
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java:119]: The change seems unnecessary, please revert it

File: hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java [hunk 3]
```diff
@@ -125,18 +129,21 @@ public Object get(int i) {
...
 
 		public <X> X get(int i, Class<X> type) {
-			final Object result = get( i );
-			if ( result != null && !type.isInstance( result ) ) {
+			final Object untyped = get( i );
+			if (type.isPrimitive() && untyped != null && !untyped.getClass().isPrimitive()) {
+				type = PrimitiveWrapperHelper.getDescriptorByPrimitiveType(type).getWrapperClass();
+			}
+			if ( untyped != null && !type.isInstance( untyped ) ) {
 				throw new IllegalArgumentException(
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java:137]:
```suggestion
			if ( result != null && !type.isInstance( result ) && ( !type.isPrimitive()  || !PrimitiveWrapperHelper.getDescriptorByPrimitiveType(type).getWrapperClass().isInstance( result ) ) {
				throw new IllegalArgumentException(
```

File: hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java [hunk 4]
```diff
@@ -125,18 +129,21 @@ public Object get(int i) {
...
 								"Requested tuple value [index=%s, realType=%s] cannot be assigned to requested type [%s]",
 								i,
-								result.getClass().getName(),
+								untyped.getClass().getName(),
 								type.getName()
 						)
 				);
 			}
-			return (X) result;
+			return type.cast(untyped);
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java:146]: The change seems unnecessary, please revert it

**Review by beikov — CHANGES_REQUESTED** [review]
Hey there, thanks for the PR. I have a few improvements though which I would like to ask you to implement. Instead of penalizing every call to `get` with additional checks, we should instead only check if a type is a primitive when `isInstance` returns false. If you could please do the changes I requested, I'd merge this.

File: hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java [hunk 5]
```diff
@@ -70,8 +71,8 @@ public <X> X get(TupleElement<X> tupleElement) {
 						"Requested tuple element did not correspond to element in the result tuple"
 				);
 			}
-			// index should be "in range" by nature of size check in ctor
-			return (X) tuples[index];
+			Class<? extends X> type = tupleElement.getJavaType();
+			return get(index, type);
```
[the remark this answers is not in the document]
Comment (karge-itestra) [reply hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java:75]: I would have to apply the same fix as in `get(index, type)` to that method. Would you prefer that I duplicate the primitive type check?

[the remark this answers is not in the document]
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java:75]: There would be something seriously wrong if the java type of the tuple element is not compatible with the objects we have in the array, so you don't need to change anything in this method.

File: hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java (hunk 2 above)
Comment (karge-itestra) [reply hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java:119]: The cast to `X` generates a warning that there is an unchecked cast, while `type.cast(untyped)` does not. The `NativeTupleImpl#get(String, Class<X>)` method does it like that as well. I can revert it but that brings back the warning.

**Comment by karge-itestra** [comment]
@beikov thanks for the suggestions, after this force-push all your requirements are satisfied, plus redundancy is reduced, and the tests are more thorough.

File: hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java:75]: Please revert this, then I can merge the PR

[the remark this answers is not in the document]
Comment (karge-itestra) [reply hibernate-core/src/main/java/org/hibernate/jpa/spi/CriteriaQueryTupleTransformer.java]: Sorry, I thought I had done this already, but it has slipped through. Now it is reverted.

**Comment by beikov** [comment]
Thanks! After CI finishes, I'll merge this

**Comment by beikov** [comment]
Looks like there is a checkstyle issue. Could you please run `./gradlew :hibernate-core:checkstyleMain` and fix the problems?

**Comment by karge-itestra** [comment]
> Looks like there is a checkstyle issue. Could you please run `./gradlew :hibernate-core:checkstyleMain` and fix the problems?

Yep, looks like spaces to fine-tune indentation is not allowed.

**Comment by karge-itestra** [comment]
The spaces for indentation detected by checkstyle have been fixed

**Comment by beikov** [comment]
Thanks for the fix!

**Comment by karge-itestra** [comment]
Shall I port the test to the current main branch and see whether that is still an issue, and make a PR for that if it is?

**Comment by beikov** [comment]
>Shall I port the test to the current main branch and see whether that is still an issue, and make a PR for that if it is?

That would be nice, yeah :)

>you´re welcome, looking forward to the next maintainance release to have it in my project ;)

I'll try to get it released by the end of this or next week. I have one more backport that I want to do before finishing this 5.6 release

**Pull request merged**